src/Prove.hs

```diff
@@ -0,0 +1,249 @@
+   - an assertion that the invariant does not hold if either of the following is true:
+      1. The constructor predicate holds
+      2. The invariant holds over the prestate and one of the method level predicates holds
+
+   If this query returns `unsat` then the invariant must hold over the transition system
+-}
+queries :: [Claim] -> [Symbolic ()]
+queries claims = fmap mkQuery $ gather claims
+
+
+-- *** Data *** --
+
+
+data When = Pre | Post
+  deriving (Show)
+
+data Ctx = Ctx Id Id When [(Id, SMType)]
+  deriving (Show)
+
+data SMType
+  = SymInteger (SBV Integer)
+  | SymBool (SBV Bool)
+  | SymBytes (SBV [WordN 8])
+  deriving (Show)
+
+catInts :: [(Id, SMType)] -> [(Id, SBV Integer)]
+catInts ((name, SymInteger i):tail) = (name, i):(catInts tail)
+catInts (_:tail) = catInts tail
+catInts [] = []
+
+catBools :: [(Id, SMType)] -> [(Id, SBV Bool)]
+catBools ((name, SymBool b):tail) = (name, b):(catBools tail)
+catBools (_:tail) = catBools tail
+catBools [] = []
+
+catBytes :: [(Id, SMType)] -> [(Id, SBV [WordN 8])]
+catBytes ((name, SymBytes b):tail) = (name, b):(catBytes tail)
+catBytes (_:tail) = catBytes tail
+catBytes [] = []
+
+
+-- *** Pipeline *** --
+
+
+-- |Builds a mapping from Invariants to a list of the Pass Behaviours for the
+-- contract referenced by that invariant.
+gather :: [Claim] -> [(Invariant, Storage, [Behaviour])]
+gather claims = fmap (\i -> (i, getStore i, getBehaviours i)) invariants
+  where
+    invariants = catInvs claims
+    getBehaviours (Invariant c _) = filter isPass $ filter (\b -> c == (_contract b)) (catBehvs claims)
+    -- TODO: refine AST so we don't need this head anymore
+    getStore (Invariant c _) = head $ filter (\(Storage n _) -> c == n) (catStores claims)
+    isPass b = (_mode b) == Pass
+
+-- |Builds a query asking for an example where the invariant does not hold.
+mkQuery :: (Invariant, Storage, [Behaviour]) -> Symbolic ()
+mkQuery (inv, store, behvs) = do
+  inits' <- mapM (mkInit inv store) inits
+  methods' <- mapM (mkMethod inv store) methods
+  constrain $ sOr (inits' <> methods')
+  where
+    inits = filter _creation behvs
+    methods = filter (not . _creation) behvs
+
+-- |Given a creation behaviour return a predicate that holds if the invariant does not
+-- hold after the constructor has run
+mkInit :: Invariant -> Storage -> Behaviour -> Symbolic (SBV Bool)
+mkInit (Invariant contract e) (Storage c1 locs) (Behaviour method _ _ c2 interface pre post stateUpdates _) = do
+  -- TODO: refine AST so we don't need this anymore
+  when (contract /= c1 || contract /= c2 || c1 /= c2) $ error "Internal error: contract mismatch"
+
+  store <- mapM (makeSymbolic (Ctx contract method Pre [])) locs
+  let ctx = Ctx contract method Pre store
+
+  inv <- symExpBool ctx e                             -- invariant holds
+  state <- mapM (\(c, u) -> fromUpdate ctx u) (updates stateUpdates) -- state updates hold
+  -- TODO: handle constructor args
+  return $ (sAnd state) .&& (sNot inv)
```

Comment:
nit: these parens are not needed I think

Reply:
hmmmm maybe, I think I'll leave them in cos it makes things a little more explicit.